Implement system call handlers for various driver subsystems

tyrantlizard77 opened the pull request:
See individual commit messages for notes.

In general, all APIs which did not involve registering callbacks that run in supervisor mode, or depend on kernel objects not ported to users mode, were added to the system call table.

This does not completely convert all driver subsystems, some are still being worked on.

The patch converts i2c, adc, pwm, ethernet, aio_comparator, counter, dma, flash, ipm, rtc, watchdog driver subsystems.

Commit:
drivers: i2c: add system call handlers

i2c actually only has two entry points into the driver,
i2c_configure and i2c_transfer. All the other APIs are derived
from these.

All derived APIs now just call i2c_transfer() with appropriate args.

The handler for i2c_transfer() needs to examine the message array
and validate all the buffers involved depending on whether we are
reading or writing to them.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: adc: add system call handlers

Straightforward conversion for adc_enable/disable.

adc_read() uses a sequence table, which points to an array
of struct adc_seq_entry, each element pointing
to memory buffers. Need to validate all of these as being readable
by the caller, and the buffers writable.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: pwm: add system call handlers

pwm_pin_set_usec now defined in terms of the other two APIs.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: aio_comparator: add system calls

aio_cmp_configure() registers a callback and is skipped.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: counter: add syscall handlers

counter_set_alarm() registers a callback and has been skipped.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: flash: add system calls

The page_layout API returns pointers to kernel memory and is not
exposed to user mode. This is fine for flash_get_page_count()
and flash_get_page_info APIs since these copy the values, but some
redesign work will be needed to get flash_page_foreach() working in
user mode since we do not want the callback running in a privileged
state.

Due to the way that (even unimplemented) system call prototypes are
generated, the definition of struct flash_pages_info needed to be
moved outside of the #ifdef.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: ipm: add system calls

ipm_register_callback() omitted since it registers a callback.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: rtc: add system calls

rtc_set_config() omitted since it registers a callback.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

lpereira commented on subsys/driver_handlers/eth.c:
```diff
@@ -0,0 +1,17 @@
+#include <syscall_handler.h>
+#include <eth.h>
+
+_SYSCALL_HANDLER(eth_send, dev, buffer, len_p)
```
Is this in anticipation for running the network layer as a user thread? I don't see much use otherwise.

andrewboie replied:
My hope is we can get to that point someday yes. But as above, I can just drop this subsystem from the PR.

lpereira commented on include/dma.h:
```diff
@@ -197,7 +197,9 @@ static inline int dma_config(struct device *dev, u32_t channel,
  * @retval Negative errno code if failure.
  */
-static inline int dma_start(struct device *dev, u32_t channel)
+__syscall int dma_start(struct device *dev, u32_t channel);
```
Do we need DMA for user threads?  Because if we do, then it's something we need to think very carefully, as this can be used to access arbitrary memory locations.

andrewboie replied:
I don't have a clear sense of what subsystems are needed for user threads.

However keep in mind that subsystem APIs cannot be used on a driver object unless that thread has been granted permission, and the dma_config() API was omitted.

However if you think it's best to just omit DMA from user mode access (or any other subsystem) I can certainly remove it.

nashif commented:
Can we check for privileged access on the device level, i.e. when doing the binding and avoid modifying the APIs with syscalls? All APIs depend on the device structure so I was wondering if we can simplify the driver APIs by just checking at the entry, if you do not have access to the device object, then you should not be able to call any APIs...

galak commented:
I think we should keep the syscall code with the drivers rather than creating a new dir structure.  So instead of subsys/driver_handlers/adc.c  -> drivers/adc/syscall.c (or something like that).  This way as any new APIs or such are done, its all kinda seen in one location (and one doesn't forget to go update that thing over in subsys/driver_handlers/

andrewboie commented:
@galak yes I can do that, it will make it simpler to have multiple driver handler PRs in play as well

andrewboie commented:
> Can we check for privileged access on the device level, i.e. when doing the binding and avoid modifying the APIs with syscalls? All APIs depend on the device structure so I was wondering if we can simplify the driver APIs by just checking at the entry, if you do not have access to the device object, then you should not be able to call any APIs...

@nashif no that couldn't possibly work. What happens when you call a UART API with a I2C driver object? How do you validate the parameters in an API-specific way, for example validate memory bufefr parameters? If the checks are done at device_get_binding time how do you do privilege elevation? You are proposing a common entry point for all driver APIs which is impossible.

lpereira commented:
A common entry point isn't impossible; it would be akin to an `ioctl()` in the Unix world.   The device API structs contains only function pointers, so we should be able to convert a "function" index to the pointer itself, given a device structure, and call the implementation function.

I'm just thinking out loud, but consider this:

```C
struct my_device_api {
    void (*set_foo)(struct my_device *dev, int foo);
    int (*get_foo)(struct my_device *dev);
};
```

We know that `sizeof(my_device_api)` is 8 bytes, since it contains two function pointers. With this, we can generate a table at compile time with a script:

```C
struct device_api_ioctls {
    int n_funcs;
};

struct device_api_ioctls[] = {
    [DEVICE_GPIO] = { ... },
    [DEVICE_I2C] = { ... },
    [DEVICE_MY_DEVICE] = { .n_funcs = 2 },
};
```

And a hypothetical `device_operation()` syscall could be something like this:

```C
SYSCALL_DEFINE(device_op, device, op, arg1, arg2, arg3, arg4, arg5)
{
      void (*impl)();
      int device_type = get_device_type(device);

      if (!validate_device(device)) {
            return -EACCESS;
      }

      if (device_api_ioctls[device_type].n_funcs > op) {
            return -EINVAL;
      }

      impl = (u8_t *)(device->api) + op * sizeof(impl);
      return impl(device, arg1, arg2, arg3, arg4, arg5);
}
```

The user side of things could be simply:

```C
int my_device_set_foo(struct device *device, int foo)
{
        return device_op(device, 0, foo);
}

int my_device_get_foo(struct device *device)
{
        return device_op(device, 1);
}
```

Obviously this would require something else to validate arguments on a case-by-case basis as well, but we can give more thought into this; it's not impossible.

andrewboie commented:
> Obviously this would require something else to validate arguments on a case-by-case basis as well

This is not a side detail, this is the whole root of the thing.
How are you going to do this?

andrewboie commented:
```
int my_device_set_foo(struct device *device, int foo)
{
        return device_op(device, 0, foo);
}

int my_device_get_foo(struct device *device)
{
        return device_op(device, 1);
}
```

I really don't like this either, we are going from having the user side completely auto-generated to having to do it by hand.

What problem are we solving here? What are you trying to get rid of? It seems you are proposing we get rid of the handlers and I maintain that is not possible.

andrewboie commented:
```
      if (device_api_ioctls[device_type].n_funcs > op) {
            return -EINVAL;
      }

      impl = (u8_t *)(device->api) + op * sizeof(impl);
      return impl(device, arg1, arg2, arg3, arg4, arg5);
```

Another way this doesn't work: this implies you can access all APIs for a subsystem from user mode. For almost all subsystems this is not true and only a subset of APIs are available to user mode.

Commit:
sensors: move userspace handlers location

Place driver handlers with the drivers, not somewhere separate under
subsys/

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: pinmux: add system call handlers

pinmux_pin_get() needs memory validated for the func parameter since
it's a pointer that gets written to.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: uart: add system call handlers

Certain interrupt-driven APIs were excluded as they are intended
only to be called from ISRs, or involve registering a callback
which runs in interrupt context.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: gpio: add system call handlers

Many APIs had two versions, by port and by pin, which called the same
API with different parameters. This has been reorganized to reduce
the number of system calls.

Callback registration API skipped.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: random: add random_get_entropy sys call

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: spi: add system call handlers

spi_transceive_async() omitted as we don't support k_poll objects
in user mode (yet).

The checking for spi_transceive() is fairly complex as we have to
validate the config struct passed in along with device instances
contained within it.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
Moved handler C files to live with the drivers per @galak suggestion

Added handlers for some more subsystems:
pinmux, uart, gpio, random, spi

Commit:
userspace: remove some driver object types

Use-cases for these  subsystems appear to be limited to board/SOC
code, network stacks, or other drivers, no need to expose to
userspace at this time. If we change our minds it's easy enough
to add them back.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: schedule_api: slightly increase stack size

Two tests were on the knife-edge of their current stack limit and
were overflowing when UART system calls were added and userspace
enabled.

Test case stack sizes are often pulled out of thin air, the current
value of 256 was just a guess.

Kick these stacks up to 384; verified with sanitycheck --all that
this doesn't break anything.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andyross commented on drivers/ipm/ipm_handlers.c:
```diff
@@ -0,0 +1,28 @@
+	_SYSCALL_MEMORY_READ(data, size);
+	return _impl_ipm_send((struct device *)dev, wait, id,
+			      (const void *)data, size);
+}
```
This is another where I think it might be best to leave it out.  The quark se IPM mechanism is very low level and there's nothing here to prevent e.g. a user process with access to this API from terribly confusing whatever protocol is being sent over that pipe.

Basically: an app that cares about protecting the system from the behavior of a process wouldn't give that process this API to begin with.

andrewboie replied:
I think this might be OK to leave in?
A user thread won't be able to use these APIs unless granted permissions on the device objects that represents the IPM channel. So a random thread won't be able to muck about with the protocol, only those threads that have been blessed with access.
I'm thinking of uses where user threads could be on the sending end of the pipe, shouldn't need to be supervisor to do that.

andyross commented on drivers/watchdog/wdt_handlers.c:
```diff
@@ -0,0 +1,27 @@
+ */
+
+#include <syscall_handler.h>
+#include <watchdog.h>
```
Note that this can allow protected processes to reboot the system.  Not sure that's really in-scope for "protection" or not.

Basically this is another spot where I don't really see a useful real-world case where you'd want to use this outside the kernel.

andrewboie replied:
i thought it may be useful to at least allow user threads to kick the dog.
but yeah I can't think of a concrete use-case, will drop this.

andyross commented on drivers/spi/spi_handlers.c:
```diff
@@ -0,0 +1,51 @@
+
+	if (tx_count) {
+		_SYSCALL_MEMORY_ARRAY_READ(tx_bufs, tx_count,
+					   sizeof(struct spi_buf));
```
This bit isn't enough, struct spi_buf contains a buf/len field that needs to be checked too, no?

andrewboie replied:
yes you're right, fixing this

andyross commented on drivers/spi/spi_handlers.c:
```diff
@@ -0,0 +1,51 @@
+
+	if (rx_count) {
+		_SYSCALL_MEMORY_ARRAY_WRITE(rx_bufs, rx_count,
+					    sizeof(struct spi_buf));
```
Ditto: need to validate the spi_buf struct itself, but also the memory it points to.

andyross reviewed (changes requested):
Found what looks like one genuine error in the SPI handler validation.  The rest is just whinery.

Full disclosure: my eyes were crossing pretty badly trying to review all the boilerplate changes.  I mostly tried to concentrate on the meaning of the APIs being ported or skipped, and the correctness of the validation code in the handlers.

andrewboie commented on drivers/ethernet/eth_handlers.c:
```diff
@@ -0,0 +1,17 @@
+#include <syscall_handler.h>
+#include <eth.h>
+
+_SYSCALL_HANDLER(eth_send, dev, buffer, len_p)
```
sounds reasonable, dropped

andrewboie commented:
removed ethernet, DMA, watchdog based on @andyross review.
Fix SPI handler checking of SPI buffers.
Hopefully made checkpatch.pl happy, it was throwing shade on some macros.

mnkp commented:
I'm not familiar with the overall concept and as such my comment concerns merely implementation:

Shouldn't we put declarations of all _impl_* functions in a private header file separate from the main one visible by the user? _impl_* functions look like implementation detail which should be hidden from the user. The same remark can be made about all *_api structures. The private header file would be included by the driver but not the application.

andyross reviewed (approved):
Can't find anything else to whine about
